rust/bridge/ffi/src/logging.rs

```diff
@@ -0,0 +1,99 @@
+//
+// Copyright 2020 Signal Messenger, LLC.
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+use libc::c_char;
+use std::ffi::CString;
+
+#[repr(C)]
+pub enum LogLevel {
+    Error = 1,
+    Warn,
+    Info,
+    Debug,
+    Trace,
+}
+
+impl From<log::Level> for LogLevel {
+    fn from(level: log::Level) -> Self {
+        use log::Level::*;
+        match level {
+            Error => Self::Error,
+            Warn => Self::Warn,
+            Info => Self::Info,
+            Debug => Self::Debug,
+            Trace => Self::Trace,
+        }
+    }
+}
+
+impl From<LogLevel> for log::Level {
+    fn from(level: LogLevel) -> Self {
+        use LogLevel::*;
+        match level {
+            Error => Self::Error,
+            Warn => Self::Warn,
+            Info => Self::Info,
+            Debug => Self::Debug,
+            Trace => Self::Trace,
+        }
+    }
+}
+
+pub type LogCallback = extern "C" fn(
+    target: *const c_char,
+    level: LogLevel,
+    file: *const c_char,
+    line: u32,
+    message: *const c_char,
+);
+
+pub type LogEnabledCallback = extern "C" fn(target: *const c_char, level: LogLevel) -> bool;
+
+pub type LogFlushCallback = extern "C" fn();
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct FfiLogger {
+    enabled: LogEnabledCallback,
+    log: LogCallback,
+    flush: LogFlushCallback,
+}
+
+impl log::Log for FfiLogger {
+    fn enabled(&self, metadata: &log::Metadata) -> bool {
+        let target = CString::new(metadata.target()).expect("no 0 bytes in log target");
+        (self.enabled)(target.as_ptr(), metadata.level().into())
+    }
+
+    fn log(&self, record: &log::Record) {
+        let target = CString::new(record.target()).expect("no 0 bytes in log target");
+        let file = record
+            .file()
+            .map(|file| CString::new(file).expect("no 0 bytes in file"));
+        let message = CString::new(record.args().to_string()).unwrap_or_else(|_| {
+            CString::new(record.args().to_string().replace("\0", "\\0")).unwrap()
+        });
+        (self.log)(
+            target.as_ptr(),
+            record.level().into(),
+            file.as_ref()
+                .map(|file| file.as_ptr())
+                .unwrap_or(std::ptr::null()),
+            record.line().unwrap_or(0),
+            message.as_ptr(),
+        );
+    }
+
+    fn flush(&self) {
+        (self.flush)()
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn signal_init_logger(max_level: LogLevel, logger: FfiLogger) {
+    log::set_logger(Box::leak(Box::new(logger))).unwrap();
```

Comment:
Can / should we detect multiple calls to this function, since each invocation causes a leak?

Reply:
`log::set_logger` will return an error if it's called multiple times, which I've turned into a panic with the `unwrap()`. For setup code like this I think that's reasonable, especially when there aren't even explicit calls to it on the client side (when using SignalCoreKit).

Reply:
Ah, I had missed that. But unwinding a Rust panic into C/Swift will cause undefined behavior. We should either return an error code or do a whole process abort if we don't like what's happening.

[My preference would be to have this fn return an int error code if this is the 2nd registration, then have the static constructor abort if the function errors, since in that context there really isn't anything else to do. Just aborting here is fine to. But UB, no bueno]

Reply:
Ugh, right, I keep forgetting that Rust panics unwind by default instead of aborting by default (which we decided is the behavior we want for this library). Thanks for catching this.